### PR TITLE
New foes for the Collapsed Tower

### PR DESCRIPTION
**Pull request by John-Candlebury**
### Summary

```SUMMARY: Content "Adds new foes to the collapsed tower"```

#### Purpose of change

More Prep Work for Hub 01 third mission. The tower was missing a thematic  boss monster in the first level, so this adds it.

#### Describe the solution

Flesh tendrils are part of some gigantic spooky undead creature that is trapped underneath the tower. they are vicious up close, and at range attack by spawning the weaker crawler foes. The main dangers are taking to long and getting fatally slowed by the crawler impalers/poison gas and then getting mobbed to death. Well if you get very unlucky in melee, the two tendrils might combo fling you for about half your health (it happened only once in about 12 tries,)
 
I tried to make a boss fight that wouldn't that difficult to beat or to escape from, but that couldn't be trivially defeated through ranged weapons. . Explosives and fire will work decently, but you have to be careful to not get caught by shrapnel, since  the room is very small and usually covered in rubble

#### Describe alternatives you've considered

I tried to handle the tendril attacks purely through json, using a combination of guns, spells and other hard-coded attacks. But this caused infrequent crashes to desktop, most commonly whenever the tendril lost sight of my player, so I went with full C++ instead. I hope that's not a problem.

### CODE REVIEW

File: data/json/monsters/monsters.json [hunk 1]
```diff
@@ -3892,12 +3892,107 @@
...
     "special_attacks": [ { "type": "leap", "cooldown": 5, "max_range": 5, "allow_no_target": true }, [ "scratch", 5 ] ],
     "death_function": [ "GAS" ],
-    "flags": [ "SEES", "HEARS", "SMELLS", "WARM", "POISON", "NO_BREATHE", "CLIMBS", "HARDTOSHOOT" ]
+    "flags": [ "SEES", "HEARS", "SMELLS", "WARM", "POISON", "CLIMBS", "NO_BREATHE", "CLIMBS", "HARDTOSHOOT" ]
+  },
+  {
+    "id": "mon_zombie_gasbag_impaler",
+    "type": "MONSTER",
+    "name": "gangrenous impaler",
+    "description": "A corpse hideosly twisted into an insectlike form.  A hollow tendril reaches out from its open thorax.",
```
Comment (Petethegoat) [inline data/json/monsters/monsters.json]:
```suggestion
    "description": "A corpse hideously twisted into an insect-like form.  A hollow tendril reaches out from its open thorax.",
```

File: data/json/items/ammo/barb.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
+  "type": "AMMO",
+  "id": "barb_paralysis",
+  "price": 400,
+  "name": "paralyzing barb",
+  "symbol": "=",
+  "color": "green",
+  "description": "Tiny droplets of venom trickle out of the sponge-like tissue of this sharp barb ",
```
Comment (Petethegoat) [inline data/json/items/ammo/barb.json]:
```suggestion
  "description": "Tiny droplets of venom trickle out of the sponge-like tissue of this sharp barb.",
```

File: data/json/items/ammo/barb.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  "material": "bone",
+  "volume": 1,
```
Comment (Petethegoat) [inline data/json/items/ammo/barb.json]:
```suggestion
  "volume": "250ml",
```

File: data/json/items/gun/monster_gun.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "WATERPROOF_GUN",
+    "NO_SALVAGE",
+    "NO_UNLOAD",
+    "RELOAD_AND_SHOOT"
+  ],
+  "skill": "pistol",
+  "ammo": "barb",
+  "clip_size": 1,
+  "weight": 540,
+  "volume": 3,
```
Comment (Petethegoat) [inline data/json/items/gun/monster_gun.json]:
```suggestion
  "volume": "750ml",
```

File: data/json/monsters/monsters.json [hunk 5]
```diff
@@ -3892,12 +3892,107 @@
...
+    "harvest": "exempt",
+    "special_attacks": [ [ "FLESH_TENDRIL", 1 ] ],
+    "death_function": [ "GAS" ],
+    "flags": [ "SEES", "HEARS", "SMELLS", "IMMOBILE", "WARM", "POISON", "IMMOBILE", "NO_BREATHE", "FILTHY" ]
+  },
+  {
+    "id": "mon_zombie_living_wall",
+    "type": "MONSTER",
+    "name": "flesh wall",
+    "description": "A great lump of mutated flesh.  Its resembles the innards of some gigantic creature and is covered in a grid of diminutive veins.",
```
Comment (Petethegoat) [inline data/json/monsters/monsters.json]:
```suggestion
    "description": "A great lump of mutated flesh.  It resembles the innards of some gigantic creature and is covered in a grid of diminutive veins.",
```

Comment (Petethegoat) [reply data/json/monsters/monsters.json]: Not sure if diminutive is the right word here either?

File: src/monattack.cpp [hunk 6]
```diff
@@ -4666,6 +4673,75 @@ bool mattack::riotbot( monster *z )
...
 }
 
+bool mattack::flesh_tendril( monster *z )
+{
+
+    Creature *target = z->attack_target();
+
+    if( target == nullptr || !z->sees( *target ) ) {
+        if( one_in( 70 ) ) {
+            add_msg( _( "The floor trembles underneath your feet" ) );
```
Comment (Petethegoat) [inline src/monattack.cpp]:
```suggestion
            add_msg( _( "The floor trembles underneath your feet." ) );
```

File: src/monattack.cpp [hunk 7]
```diff
@@ -4666,6 +4673,75 @@ bool mattack::riotbot( monster *z )
...
+        }
+        if( !free.empty() ) {
+
+            z->moves -= 100;
+            const tripoint target = random_entry( free );
+            if( monster *const summoned = g->summon_mon( spawned, target ) ) {
+                summoned->make_ally( *z );
+                g->m.propagate_field( z->pos(), fd_gibs_flesh, 75, 1 );
+                if( g->u.sees( *z ) ) {
+                    add_msg( m_warning, _( "A %s stuggles itself free from the %s!" ), summoned->name(),
```
Comment (Petethegoat) [inline src/monattack.cpp]:
```suggestion
                    add_msg( m_warning, _( "A %s struggles to pull itself free from the %s!" ), summoned->name(),
```

**Comment by Night-Pryanik** [comment]
Honestly this look like a 100% fantasy monster to me. If it's a mutant, then what was the base for it? The one that a) has ability to spawn poisonous paralyzing barbs; b) has ability to spawn other creatures out of its body; c) has immense size; d) has some poisonous gas inside that explodes on its death. And other questions.
I think this monstrosity is a pure fiction and couldn't exist even in the Cataverse.

**Comment by DemAvalon** [comment]
Yes because Jonh Carpenter's the Thing, noodle horror man, mi-go's, walking plants, funguys, shoggoths, and floating eyeballs are super realistic.

**Comment by Night-Pryanik** [comment]
As it was said about of million times, having one questionable content in a game isn't a reason to add another questionable content.

**Comment by DemAvalon** [comment]
As it was pointed out multiple times, the whole game's premise is questionable.

**Comment by Night-Pryanik** [comment]
So instead of fixing it, let's worsen it by adding more questionable content?

**Comment by DemAvalon** [comment]
To "fix it" you would have to remove zombies from this zombie game. 
also, what is so unrealistic about this monster? it's like the Jabberwock but on a much larger scale and with paralyzing barbs, the goop seems to have the ability to combine multiple "paths" of evolution (skeletal shocker) why would it not be able to do this?

**Comment by Night-Pryanik** [comment]
> To "fix it" you would have to remove zombies from this zombie game.

Having a fantastic element in an otherwise realistic element doesn't require fixing.

> the goop seems to have the ability to combine multiple "paths" of evolution (skeletal shocker)

Skeletal shocker is not an amalgam of several mutation "paths", shockers ain't mutants, they are "ordinary" zombies, just with installed bionics.

> why would it not be able to do this?

Because there are too much non-crossing mutations for one monster. How it was created in first place? Scientists injected some "base" creature (once again, what it was?) all mutagens they had at hand?

**Comment by DemAvalon** [comment]
It seems to be a bunch of things that just got fused together, many mutants/people ala Jabberwock. the shocker zombie is a mutation on the zombie otherwise every zombie that has CBM would be shockers (scientist, bio operator, technicians)

> Having a fantastic element in an otherwise realistic element doesn't require fixing.

what is the realistic element in scientists on an alternate earth opening portals to other dimensions and finding an extra-dimensional black and less tasty version of "The Stuff" that can open portals to other dimensions on a whim who also for some reason hijacked corpses and causes them to mutate in other to carry its goals?

**Comment by Night-Pryanik** [comment]
> every zombie that has CBM would be shockers (scientist, bio operator, technicians)

No it's not. Only those zombies who have some kind of electricity-related CBM become shockers.

> portals to other dimensions and [...] an extra-dimensional black and less tasty version of "The Stuff"

**THIS** is a fantastic element in otherwise realistic universe.

**Comment by DemAvalon** [comment]
> No it's not. Only those zombies who have some kind of electricity-related CBM become shockers.

Like power storages? the thing the vast majority of CBMs need to work? including the combat CBM that the operators use and the electromagnetic CBM that the Technicians use.

> THIS is a fantastic element in otherwise realistic universe.

Then how is this monster (the one on the PR) any more objectionable? it's also a fantastic element in an otherwise "realistic" universe.

**Comment by Night-Pryanik** [comment]
> Like power storages?

No, like Chain Lightning CBM.

> Then how is this monster (the one on the PR) any more objectionable?

Having one fantastic element (portals and blob from another dimension, which leads to mutations and zombies) is OK. 
Having another fantastic element not directly related to the first element (mycus and other netherworld creatures like mi-go, gozu, kreck and the like from another dimension) is still OK, but begin to strain the suspension of disbelief. 
Having third (and so on) fantastic element not directly related to the first element (amigara, dark wyrms, the Thing and the like) is making the game more and more unbelievable. 
The more fantastic elements are added, the more the whole setting loses its consistency, plausibility and believability.

So we shouldn't add new elements just for the sake of new elements unless they have a strong internally consistent rationale. And I personally don't see a good reason for this monster to exist at all except for being a fantastic boss at the end of the the dungeon.

**Comment by DemAvalon** [comment]
But all mutants are connected directly to the black jello are they not? since that is what causes the mutations, so this monster is directly related to the first thing, personally I think adding elements that are unexplainable adds to the mystery and wonder of the game rather than taking away from it but that is also what my opinion is.

I also want to make it very clear that though I disagree with your opinion, I still respect you as an individual and it is not my wish to offend you in any way, shape or form.

**Comment by John-Candlebury** [comment]
DemAvalon is right in here, the monster works exactly like a Jabberwock.   In this case it the monster came to be from the human remains that were trapped in the debris, which were joined together into a sensless mass of living flesh by the blob. That is the base of it.

**Comment by BaumgeistOne** [comment]
Chipping in my two cents:

I think it's alright for the player not to know how a monster came to be, but the developers should have a proper lore explanation for the fictional stuff they put into the game. A consistent lore is something I very much appreciate in any game.

So if the explanation is "a lot of bodies got crushed in a pile, the blob merged them and created a huge monster", then I, as a player, am fine with it - but I obviously don't call the shots around here.

**Comment by Destragon** [comment]
I think it's fine for a quest boss monster to be some sort of unique creature.
I agree that consistency is important, but when the game has a plot device like the blob, which can cause all sorts of crazy, random mutations, it's kinda difficult to manage to make something that really is inconsistent.

**Comment by justamblingalong** [comment]
> So we shouldn't add new elements just for the sake of new elements unless they have a strong internally consistent rationale. And I personally don't see a good reason for this monster to exist at all except for being a fantastic boss at the end of the the dungeon.

FWIW, I find this pretty persuasive as, for example, a strong reason not to add an alien invasion; being able to trace everything back to one (or two) specific breaks from realism seems like a good rule of thumb. I'm just not sure I agree that it applies to this specific instance; the blob fusing a bunch of biomass together is already a known capability.

Where I do have a qualm is with the idea of 'boss fights;' if a monster is getting added to the game, I'd rather see it be something that can spawn in a range of situations. Having a specific 'boss monster' feels really gamey in a way that CDDA generally avoids, to its benefit.

Incidentally, this is slightly off topic, but I'd really like to see this logic applied to existing content that directly references other fiction; the Amigara Horror event, in particular, really seems like it belongs in Crazy Cataclysm, not the base game.

**Comment by Night-Pryanik** [comment]
> the monster came to be from the human remains that were trapped in the debris, which were joined together into a sensless mass of living flesh by the blob. That is the base of it.

Ok, so it's just a mess of ordinary human corpses, right? Then where did it got its extraordinary abilities from? Like paralyzing barbs and explosive gas inside.

**Comment by DemAvalon** [comment]
> Ok, so it's just a mess of ordinary human corpses, right? Then where did it got its extraordinary abilities from? Like paralyzing barbs and explosive gas inside.

Same place the zombies get them I imagine.

**Comment by John-Candlebury** [comment]
>Ok, so it's just a mess of ordinary human corpses, right? Then where did it got its extraordinary abilities from? Like paralyzing barbs and explosive gas inside.

If you want the rough biological processes that inspire it: they are gas gangrene, and tick paralysis, barbs are a bone deformation thing, like all the bone related stuff that already happens.  Obviously they are exaggerated into mutations that will affect game play, just like with other zombie abilities like spitting acid, sniping you with acid, exploding into gas, and vomiting glowing bile. For all I know this fits with the established abilities of the blob, and in fact, most of the abilities of the monster, bar the spawning, already exist as zombie abilities. 

Honestly, as I read it, your argument just comes down as unconstructive criticism of the" "This doesn't mach my preconceived vision of the game, so you shouldn't have done it" variety. Which is an opinion I obviously dont agree with, since I already spent a pair of days building the dungeon, and if you excuse me it is also an opinion I dont want to engage with any further.

So I'll just leave this like it is, and wait for something else to happen. As always thank you for your technical code reviews.

**Comment by nexusmrsep** [comment]
I have to unfortunately agree with @NightPryanic in this particular case. While a a Jabberwock would be plausible for a pile of corpses merged together with the blob, the rest is not. As observed in game zombie mutations come from a reason and for a reason, often utilizing some pre-existing condition of the cadaver. Thus distinct mutation lines. Strong corpse, strong zombies. Corpse of a swimmer, swimmer zombie. Acid in stomach, let's make a weapon from it, blob says. Etc. And blob builds on that on later iterations. Maybe it's not yet fleshed out in the lore, but everything seems on purpose. There is a cause, there is effect, there is purpose. 
What's the cause of these special abilities and qualities? From where do they come from? Where do they lead? What blob wants to achieve here? It looks odd, not in place, it seems off and feels off. You look at it and first thing is asking yourself why?

And it's required. Consistency of such things is critical to building any fantastic world. So @DemAvalon - you are as far off the point as one can get. A good fantastic world is a consistent fantastic world. Not a bucket full of things thrown in for no reason just for a sake of it. And realism has nothing to do with it. Or mixing conventions. You can have elves hunting aliens if you keep it consistent to the story and sell it the right way. Things that are real are real. Things that are not must keep to the design and be consistent. Drop the logic and the image crumbles. At any point player should not doubt that pieces don't fit together.

**Comment by DemAvalon** [comment]
I cannot fathom why a real big Jabberwock with barbs and poison is such a difficult concept to grasp.

Goop saw a bunch of corpses mushed under a collapsed tower, it said "I wanna get in on that action" since bodies where mushed from debris it though "Yo what if... I merge them all together, that sounds rad imma do it!" so it did, the thing has barbs because the black "Stuff" wants it to have, it has gas because it wants the creature to have it.

This amorphous extra-dimensional wizard can alter the very fabric of our universe on a whim, is supposedly intelligent but it needs the bodies of the most intelligent people on earth to get enough brainpower to figure out how to open a door, and this monster specifically is what causes you to suspend your disbelief, a giant thorny Jabberwok loaded with fart gas is too much for your extra-dimensional god monster invading and wrecking earth story.

You have the first interesting creature in a long time that is not a robot or a variant of something else and you are fighting against adding it, for what? what would you replace it with? another big zombie?

Also @nexusmrsep : "_Yeah, well, that's just, like, your opinion, man._"
                                                                                                   -**The Dude**. 1998.

File: src/monattack.cpp [hunk 8]
```diff
@@ -4666,6 +4673,75 @@ bool mattack::riotbot( monster *z )
...
+                    add_msg( m_warning, _( "A %s struggles to pull itself free from the %s!" ), summoned->name(),
+                             z->name() );
+                }
+            }
+            return true;
+        }
+    }
+
+    if( 2 <= distance_to_target && distance_to_target <= 3 && one_in( 4 ) ) {
+        //it  it pulls you to then knock you away
```
Comment (Night-Pryanik) [inline src/monattack.cpp]:
```suggestion
        // it pulls you to it then knocks you away
```

File: src/monattack.cpp [hunk 9]
```diff
@@ -4666,6 +4673,75 @@ bool mattack::riotbot( monster *z )
...
+                if( g->u.sees( *z ) ) {
+                    add_msg( m_warning, _( "A %s struggles to pull itself free from the %s!" ), summoned->name(),
+                             z->name() );
+                }
+            }
+            return true;
+        }
+    }
+
+    if( 2 <= distance_to_target && distance_to_target <= 3 && one_in( 4 ) ) {
```
Comment (Night-Pryanik) [inline src/monattack.cpp]: Distance to target is more than or equals to 2 AND distance to target is less than or equals to 3? Will it ever be true?

Comment (AMurkin) [reply src/monattack.cpp]: Yes, distance_to_target equal to 2 or 3.

Comment (Night-Pryanik) [reply src/monattack.cpp]: Then why don't write it like `distance_to_target == 2 || distance_to_target  == 3`?

File: src/creature.cpp [hunk 10]
```diff
@@ -718,6 +720,10 @@ void Creature::deal_projectile_attack( Creature *source, dealt_projectile_attack
     if( proj_effects.count( "APPLY_SAP" ) ) {
         add_effect( effect_sap, 1_turns * dealt_dam.total_damage() );
     }
+    if( proj_effects.count( "PARALYZEPOISON" ) &&  dealt_dam.total_damage() > 0 ) {
+        add_msg_if_player( m_bad, _( "You feel poison enter your body!" ) );
```
[the remark this answers is not in the document]
Comment (John-Candlebury) [reply src/creature.cpp]: I think the correct spelling is  actually "enter"

File: data/json/monsters/monsters.json [hunk 11]
```diff
@@ -3892,12 +3892,107 @@
...
+    "melee_cut": 0,
+    "vision_night": 30,
+    "armor_bash": 6,
+    "armor_cut": 6,
+    "harvest": "exempt",
+    "starting_ammo": { "barb_paralysis": 100 },
+    "special_attacks": [
+      {
+        "type": "gun",
+        "cooldown": 2,
```
[the remark this answers is not in the document]
Comment (John-Candlebury) [reply data/json/monsters/monsters.json]: Not really, between the monster choosing to leap, and its very slow speed, the attack triggers about once every 15 turns.

File: src/creature.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Petethegoat) [reply src/creature.cpp]: "Enter" is correct here, yeah. Although I'd say it doesn't read particularly well either way.
Maybe something like
```suggestion
        add_msg_if_player( m_bad, _( "You feel poison coursing through your body!" ) );
```

**Comment by jbytheway** [comment]
> I tried to handle the tendril attacks purely through json, using a combination of guns, spells and other hard-coded attacks. But this caused infrequent crashes to desktop, most commonly whenever the tendril lost sight of my player

It would be good to understand the causes of these crashes if you can make a reproducible example and post an issue (perhaps you already have, but I didn't see mention of it here).

**Pull request merged**